File: python/ex3/ex3.py
import numpy as np
import scipy.io as scio
import matplotlib.pyplot as plt

# 需要自己写的模块
from displayData import display_data
from lrCostFunction import lr_cost_function
from oneVsAll import one_vs_all
from predict import predict_one_vs_all
# 使用Ligistic Regression进行多分类

# ============================== 1.读取数据和初始化 ============================
# 使用scipy.io中的函数读取mat文件，data的格式是字典
data = scio.loadmat('ex3data1.mat')
# 根据关键字，分别获得输入数据和输出的真值
# print(type(Y),type(X)) # X和Y都是numpy.narray格式，也就是数组格式
X = data['X']
Y = data['y']
print(X.shape)

# 随机取出其中的100个样本，显示结果
m = X.shape[0]
# 从[0,m-1]之间，随机生成一个序列
rand_indices = np.random.permutation(range(m))
selected = X[rand_indices[1:100],:]
# 显示手写数字样例
display_data(selected)

# ========================= 2.向量化Logistic Rgression =========================
# 测试函数lr_cost_function的功能
"""
theta_t = np.array([-2, -1, 1, 2])
X_t = np.c_[np.ones(5), np.arange(1, 16).reshape((3, 5)).T/10]
y_t = np.array([1, 0, 1, 0, 1])
lmda_t = 3
cost,grad = lr_cost_function(X_t,y_t,theta_t,lmda_t)
np.set_printoptions(formatter={'float': '{: 0.6f}'.format})
print('Cost: {:0.7f}'.format(cost))
print('Expected cost: 3.734819')
print('Gradients:\n{}'.format(grad))
print('Expected gradients:\n[ 0.146561 -0.548558 0.724722 1.398003]')
"""
# 训练模型

lmd = 0.01
num_labels = 10
all_theta = one_vs_all(X,Y,num_labels,lmd)

# =============================== 3.预测 =======================================
pred = predict_one_vs_all(X,all_theta)
# 这里一定要把Y.shape变成(m,)，否则Y.shape = (m,1)，带入是无效的
Y = Y.reshape(Y.size)
print('Training set accurayc:{}'.format(np.mean(pred == Y)*100))

File: matlab/ex3/ex3.m
%% Machine Laearning Online Class - Exercise 3: One-vs-All Logistic Regression
% 多个Logistic Regression同时运算，用来对手写数字进行识别
close all;clc
%% 1.读取数据和初始化，
fprintf('Load dataSet:ex3data1.mat\n');
load('ex3data1.mat');
% 从数据中随机选择100个样本，显示成10*10的数字表格
randIndex = randperm(size(X,1));
seldata = X(randIndex(1:100),:);
% 数据可视化
figure(1);
displayData(seldata);

%% 2.向量化Logistic Regression，并计算代价函数和梯度值
fprintf('Training data with one-Vs-All logistic regression\n');
% 设置正则化系数lambda，和输出的样本类别数labels
lambda = 0.05;
num_labels=10;
[all_theta] = oneVsAll(X,y,num_labels,lambda);

%% 3.使用OneVsAll来预测
P=predictOneVsAll(X,all_theta);
fprintf('Training set Accuracy is %f\n', mean(double(P==y)) * 100);
